### PR TITLE
[21231] apply table styling

### PR DESCRIPTION
**Pull request by HDinger**
started with the 'Admin/Users' table

https://community.openproject.org/work_packages/21231

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://hdinger-feature-21231-apply-table-styling.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://hdinger-feature-21231-apply-table-styling.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by HDinger** [comment]
@ulferts can you please have a look at the 'sort_header_tag' and it's functionality? 
Thanks a lot!

**Comment by oliverguenther** [comment]
@HDinger @ulferts I've encountered and addressed the issue with the LSG sort_header_tag in my own PR. This currently adds a second helper method to provide a column for LSG table headers, and the other should be removed once all tables are re-styled.

Maybe we should merge into a separate branch beside dev to have early access to the issues we both have to address (changed sort headers, re-working pagination style, ..)

https://github.com/oliverguenther/openproject/blob/feature/projects_pagination_and_sort/app/helpers/sort_helper.rb#L282-L291

**Comment by ulferts** [comment]
I have now adapted the helper in https://github.com/HDinger/openproject/pull/1. As for the integration branch, I fear that the added complexity is too much. We simply need to get this feature into dev.

**Comment by oliverguenther** [comment]
https://github.com/HDinger/openproject/pull/2 should fix the broken cukes.

**Comment by ulferts** [comment]
I addressed the problems you mentioned with regards to table headers on tabs not being their desired width in https://github.com/HDinger/openproject/pull/4 @HDinger

**Comment by oliverguenther** [comment]
With the new styling, the empty headers in the admin info table looks weird due the additional line from the System caption.
bildschirmfoto 2015-08-14 um 14 52 41

In the plugins table, that line is also visible, but there is no line above it, it doesn't bother me so much.

Not sure If you agree, but if you do, I'd suggest either:
1. Add headers
2. Allow the header section to be hidden explicitly.

**Comment by oliverguenther** [comment]
In the auth sources table, it's the other way around. Headers are visible, but no content.

What do you think of the following empty row notice?
bildschirmfoto 2015-08-14 um 15 01 14

**Comment by HDinger** [comment]
In my eyes the missing header doesn't look weird. @ulferts what do you think?
I like the solution for the case of missing content. For me this is a better way for missing content than the big notification that is displayed in all other tables.

**Comment by oliverguenther** [comment]
On the restyled forums table, topics with no reply yet have a one-line row, while others with replies have two lines. This is low priority, but one could think about making the two-line by default, so it looks homogenous in most cases.

bildschirmfoto_2015-08-14_um_16_06_37

**Comment by oliverguenther** [comment]
On org-config: 
1. Roles and permissions are overridden by `openproject-global_roles/app/views/roles/index.html.erb` and receive the old styling
2. The projects/settings/member table is modified by two hooks in `openproject-costs/app/views/hooks/costs/_view_projects_settings_members_table_{header,row}.html.erb` , which results in a wrong header placement.

bildschirmfoto 2015-08-14 um 16 46 08

**Comment by HDinger** [comment]
I can take care of the first problem.
The second problem is already fixed in an extra PR in the budget-plugin

https://github.com/finnlabs/openproject-costs/pull/153

**Comment by ulferts** [comment]
@HDinger the check/uncheck all functionality no longer works (project administration -> types):
image

**Comment by ulferts** [comment]
> In my eyes the missing header doesn't look weird. @ulferts what do you think?

I don't miss the headers. However, I think that it is wrong to use a table here (but it used to be this way already). The table is only taken for layouting IMO. We'd need to have a list similar to what we have on the project list:

image

However, I believe this is not necessarily within the scope of this PR.

**Comment by ulferts** [comment]
> What do you think of the following empty row notice?

I believe we should stick to the large notice because it is used everywhere. Additionally, I would find it weird to have the information that the table holds no information as part of the table. Thinking of screen readers, I would be wondering, how they would read the information then. It would probably read "Name" > "Es wurden keine Authentifizierungs-Arten hinterlegt".

**Comment by oliverguenther** [comment]
You're right, the big notice is favorable. The notification was a quick idea using a colspan, but I admit that I do not know what screenreaders would output for a row with colspan over the whole table.

**Comment by HDinger** [comment]
I fixed the 'Roles and Permissions'-Table in: https://github.com/finnlabs/openproject-global_roles/pull/41

**Comment by oliverguenther** [comment]
> The table is only taken for layouting IMO. We'd need to have a list similar to what we have on the project list
> 
> However, I believe this is not necessarily within the scope of this PR.

I agree that this is out of scope. I'll refactor it into the attribute group when I extend the admin/information section for  [Feature #21275: Show remaining available storage information](https://community.openproject.org/work_packages/21275).

**Comment by ulferts** [comment]
@HDinger I believe that there are still some styles on the repository table making it look slightly different (right align, text smaller):
image
I think those rules can simply be removed.

File: app/views/admin/plugins.html.erb [hunk 1]
```diff
@@ -29,19 +29,42 @@ See doc/COPYRIGHT.rdoc for more details.
...
+      <div class="generic-table--header-background"></div>
+    </div>
+  </div>
 <% else %>
-  <p class="nodata"><%= l(:label_no_data) %></p>
+  <div class="generic-table--container">
+    <div class="generic-table--no-results-container">
+      <h2 class="generic-table--no-results-title">
+        <i class="icon-info"></i>
+        Nothing to display
```
Comment (ulferts) [inline app/views/admin/plugins.html.erb]: Those strings need to be internationalized (please also see the other tables). That means:
- adding a key to `config/en.yml` 
- using the key here: <%= l(:[Key]) %>

File: app/views/custom_fields/_index.html.erb [hunk 2]
```diff
@@ -26,36 +26,100 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
...
+            <% if tab[:name] == 'WorkPackageCustomField' %>
+              <td><%= checked_image custom_field.is_for_all? %></td>
+              <td><%= l(:label_x_projects, :count => custom_field.projects.count) if custom_field.is_a? WorkPackageCustomField and !custom_field.is_for_all? %></td>
+            <% end %>
+            <td><%= reorder_links('custom_field', {:action => 'update', :id => custom_field}, method: :put) %></td>
+            <td class="buttons">
+              <%= delete_link custom_field_path(custom_field) %>
+            </td>
+          </tr>
+        <% end; reset_cycle %>
```
Comment (ulferts) [inline app/views/custom_fields/_index.html.erb]: the `reset_cycle` call is necessary only in combination with `cycle("odd", "even")` used for the colouring the table rows. As we don't do this any more, the call can be removed.

File: app/views/enumerations/index.html.erb [hunk 3]
```diff
@@ -32,31 +32,75 @@ See doc/COPYRIGHT.rdoc for more details.
...
+                    :confirm => l(:text_are_you_sure),
+                    :class => 'icon icon-delete' %>
+              </td>
+            </tr>
+          <% end %>
+        </table>
+        <div class="generic-table--header-background"></div>
+      </div>
+    </div>
     <% reset_cycle %>
```
Comment (ulferts) [inline app/views/enumerations/index.html.erb]: please see above

File: app/views/groups/_memberships.html.erb [hunk 4]
```diff
@@ -32,47 +32,82 @@ See doc/COPYRIGHT.rdoc for more details.
...
+                  <td class="buttons">
+                    <%= link_to_function l(:button_edit), "$('member-#{membership.id}-roles').hide(); $('member-#{membership.id}-roles-form').show(); return false;", :class => 'icon icon-edit' %>
+                    <%= link_to l(:button_delete), membership_of_group_path(@group, membership),
+                                         :method => :delete,
+                                         :remote => true,
+                                         :class => 'icon icon-delete' %>
+                  </td>
+                </tr>
+              </tbody>
+            <% end; reset_cycle %>
```
Comment (ulferts) [inline app/views/groups/_memberships.html.erb]: please see above

File: app/views/workflows/index.html.erb [hunk 5]
```diff
@@ -32,29 +32,47 @@ See doc/COPYRIGHT.rdoc for more details.
   <p class="nodata"><%= l(:label_no_data) %></p>
```
Comment (ulferts) [inline app/views/workflows/index.html.erb]: I believe this needs the proper styling as well

File: app/views/workflows/index.html.erb [hunk 6]
```diff
@@ -32,29 +32,47 @@ See doc/COPYRIGHT.rdoc for more details.
...
+              <% end %>
+            </tr>
+          </thead>
+          <tbody>
+            <% @workflow_counts.each do |type, roles| -%>
+              <tr>
+                <td><%= h type %></td>
+                <% roles.each do |role, count| -%>
+                  <td>
+                    <%= link_to((count > 0 ? count : image_tag('webalys/red_x.png')), {:action => 'edit', :role_id => role, :type_id => type}, :title => l(:button_edit)) %>
```
Comment (ulferts) [inline app/views/workflows/index.html.erb:75]: Not directly related to the PR (maybe do it in a separate one) but can we get rid of `image_tag('webalys/red_x.png')`. That way we can sever the connection to app/assets/images/webalys even more. 

I created a work package for it: https://community.openproject.org/work_packages/21312

File: app/views/workflows/_form.html.erb [hunk 7]
```diff
@@ -26,48 +26,72 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
...
+                :class => "old-status-#{old_status.id} new-status-#{new_status.id}" %>
+              </td>
+            <% end -%>
+          </tr>
+        <% end %>
+      </tbody>
+    </table>
+    <div class="generic-table--header-background"></div>
+  </div>
+</div>
```
Comment (ulferts) [inline app/views/workflows/_form.html.erb:96]: Any chance we can get a different styling for this one?

It is currently looking like this:
image

I'd rather have the status headings (checkmark + name of status) more visible as headings. To that end I would like:
- to have the column highlighting work on the the status heading
- to have the status heading in bold and moved below the stronger line

Don't know if it is possible though.

**Comment by ulferts** [comment]
Looks awesome @HDinger. I think we can merge this soon and I am really looking forward to it.

**Comment by ulferts** [comment]
@HDinger can you please try to again merge in dev now that #3397 is merged?

**Pull request merged**